### PR TITLE
add new max channel size config option

### PR DESCRIPTION
**Pull request by calvinrzachman**
- let users specify their MAXIMUM WUMBO with new config option which sets the maximum channel size lnd will accept
- current implementation is a simple check by the fundingManager rather than anything to do with the ChannelAcceptor
- Add test cases which verify that maximum channel limit is respected for wumbo/non-wumbo channels
- use --maxchansize 0 value to distinguish set/unset config. If user sets max value to 0 it will not do anything as 0 is currently used to indicate to the funding manager that the limit should not be enforced. This seems justifiable since --maxchansize=0 doesn't seem to make sense at first glance.
- add integration test case to ensure that config parsing and validation is proper. I simplified the funding managers check electing to rely on config.go to correctly parse and set up either i) non wumbo default limit of 0.16 BTC OR ii) wumbo default soft limit of 10 BTC OR iii) a user specified max channel size

Addresses: https://github.com/lightningnetwork/lnd/issues/4557

#### Pull Request Checklist

- [ ] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [ ] All changes are Go version 1.12 compliant
- [ ] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [x] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [x] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [x] Any new logging statements use an appropriate subsystem and
  logging level
- [x] Code has been formatted with `go fmt`
- [x] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [ ] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [x] Running `make check` does not fail any tests
- [x] Running `go vet` does not report any issues
- [x] Running `make lint` does not report any **new** issues that did not
  already exist
- [ ] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [ ] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

### CODE REVIEW

File: fundingmanager.go [hunk 1]
```diff
@@ -1254,7 +1259,7 @@ func (f *fundingManager) handleFundingOpen(fmsg *fundingOpenMsg) {
 	// We'll reject any request to create a channel that's above the
 	// current soft-limit for channel size, but only if we're rejecting all
 	// wumbo channel initiations.
-	if f.cfg.NoWumboChans && msg.FundingAmount > MaxFundingAmount {
+	if f.cfg.NoWumboChans && amt > MaxFundingAmount {
```
Comment (Roasbeef) [inline fundingmanager.go:1262]: 👍

**Comment by Roasbeef** [comment]
NP on not using a channel acceptor, this is much simpler after all, in the future we may opt to roll it out into one, just so we can have all the logic coupled in a single place. 

>  Should this configurable maximum be above the old soft-limit or should we permit any value? If above soft limit only, ensure that this setting does not override wumbo channel enablement (ie: don't impose a cap that is lower than the old limit)

What do you mean above the old soft limit? You mean like ensuring if wumbo _isn't_ activated, then we don't allow this value to be set? 

One other thing I think we should consider is adding a new default soft-limit if wumbo _is_ activated. Just to make sure users are protected somewhat, and need to _explicitly_ allow mega-wumbo channels.

**Comment by calvinrzachman** [comment]
Hi @roasbeef, thanks for the review. I was hoping to get the test case and validation added in before you made a first pass, but alas. I’ve added those in and made a couple changes.

At first glance I didn’t like the idea of a default setting for `--maxchansize` overriding the wumbo config option which led me to where things are currently: 

- ANY explicit user setting of ` --maxchansize` is respected (both for wumbo and non-wumbo).
- For non-wumbo configurations this limit remains `16777215` satoshis by default as specified in BOLT-0002. NOTE: This is already enforced by [this](https://github.com/lightningnetwork/lnd/blob/ef69537c7f8d7b702f585831df8e3a746b8a756b/fundingmanager.go#L1257) existing code.
- For wumbo funding managers this limit is NOT enforced by default (currently indicated by zero value). We could modify this to impose a new new default soft-limit in the wumbo case as you mentioned.

Maybe there is a better way to do this than using a zero value. This breaks down should the user set `--maxchansize=0` but perhaps it's okay since that is most certainly not a valid setting anyway. An alternative way to distinguish between the zero value and an unset value is with pointers.

One more thing: Right now I only enforce `--maxchansize` inside `handleFundingOpen()` which I believe is in response to an OpenChannel message from a peer.  It makes sense to me to enforce this on channels created by user inside handleInitFundingMsg(). Let me know what you think and whether I took this in the wrong direction.

File: config.go [hunk 2]
```diff
@@ -379,6 +380,7 @@ func DefaultConfig() Config {
 		Alias:                         defaultAlias,
 		Color:                         defaultColor,
 		MinChanSize:                   int64(minChanFundingSize),
+		MaxChanSize:                   int64(MaxFundingAmount),
```
[the remark this answers is not in the document]
Comment (calvinrzachman) [reply config.go]: ✅ added

**Review by halseth — COMMENTED** [review]
Thanks for the PR, awesome change 👍 

I agree that perhaps a soft-limit by default is preferred. If you opt into wumbo, then you should be able to open channels of any size you like, but IMO it should be a second opt-in to start accepting huuuge channels.

**Comment by calvinrzachman** [comment]
At first a new larger default seemed to defeat the nature of WUMBO channels, but given that people will be entrusting their wealth to `lnd`, that's starting to sound like a better and better idea. 

Any thoughts on what the new default maximum channel size should be?

**Comment by Roasbeef** [comment]
> Any thoughts on what the new default maximum channel size should be?



10 BTC?

**Comment by Roasbeef** [comment]
>  If you opt into wumbo, then you should be able to open channels of any size you like, but IMO it should be a second opt-in to start accepting huuuge channels.

Yeh there's a distinction here:
  * pre wumbo .16 is the limit
  * post wumbo you can open any size you want 
  * post wumbo there's a new default soft limit for _accepting_
  * post wumbo, users can modify the default _acceptance_ max

File: config.go [hunk 3]
```diff
@@ -388,7 +388,7 @@ func DefaultConfig() Config {
 		Alias:                         defaultAlias,
 		Color:                         defaultColor,
 		MinChanSize:                   int64(minChanFundingSize),
-		MaxChanSize:                   int64(MaxFundingAmount),
+		MaxChanSize:                   int64(0), // Do not enforce maxiumum channel size by default
```
Comment (Roasbeef) [inline config.go:391]: I think we _do_ want to enforce it by default, but only when we're in wumbo mode. A default wumbo soft-cap of 10 BTC has been suggested.

File: config.go [hunk 4]
```diff
@@ -624,6 +624,31 @@ func ValidateConfig(cfg Config, usageMessage string) (*Config, error) {
...
+		if cfg.MaxChanSize < cfg.MinChanSize {
+			return nil, fmt.Errorf("invalid channel size parameters: "+
+				"max channel size %v, must be no less than min chan size %v",
+				cfg.MaxChanSize, cfg.MinChanSize,
+			)
+		}
+
+		// Don't allow superflous --maxchansize greater than
+		// BOLT 02 soft-limit for non-wumbo channel
+		if !cfg.ProtocolOptions.Wumbo() && cfg.MaxChanSize > int64(MaxFundingAmount) {
```
Comment (Roasbeef) [inline config.go:644]: 👍

File: config.go [hunk 5]
```diff
@@ -624,6 +624,31 @@ func ValidateConfig(cfg Config, usageMessage string) (*Config, error) {
...
+			return nil, fmt.Errorf("invalid channel size parameters: "+
+				"maximum channel size %v is greater than maximum non-wumbo channel size %v",
```
Comment (Roasbeef) [inline config.go:646]: Should be cut to reside near the line limit.

File: fundingmanager.go [hunk 6]
```diff
@@ -1280,7 +1280,9 @@ func (f *fundingManager) handleFundingOpen(fmsg *fundingOpenMsg) {
 
 	// We'll, also ensure that the remote party isn't attempting to propose
 	// a channel that's above our current maximum channel size.
-	if amt > f.cfg.MaxChanSize {
+	// Respect the maxiumum channel size if explicitly set. If unset
+	// (marked by 0 value), then skip this check and permit channel creation of any size
```
Comment (Roasbeef) [inline fundingmanager.go:1284]: Style nit: sentence should end in a period, and also be wrapped to 80 chars.

**Review by Roasbeef — CHANGES_REQUESTED** [review]
This PR is super close! IMO we want to have a default soft-limit for for wumbo chans in place. In the past, we threw around a value of 10 BTC which seemed reasonable at the time. This would only be for _accepting_ the channels as mentioned. 

We'd like to tag our v0.11.1 version in a few days here, and I think this is really close, so I'd like to get it in.

**Comment by calvinrzachman** [comment]
Latest update contains enforcement for `10 BTC` default limit. I elected to setup defaults in `ValidateConfig()` inside `config.go` and keep the simplest check I could think of in the Funding Manager. Since the funding manager now trusts its config in determining the `MaxChanSize`, I added a new `itest` to verify that config.go handles explicit user setting of `--maxchansize` and both wumbo/non-wumbo defaults. 

I agree this is close. One thing I am unsure on right now is whether my setting of the default limits will screw up the case where lnd is run with Litecoin.  I will think on this a bit more and do my best to finish this PR off this weekend.

**Comment by Roasbeef** [comment]
> One thing I am unsure on right now is whether my setting of the default limits will screw up the case where lnd is run with Litecoin.

The default can always be overridden in that case. FWIW, the 0.16 number itself was very Bitcoin specific.

**Comment by Roasbeef** [comment]
The latest set of changes is looking pretty good to me! Final thing is to either squashed down these commits into one, or update the title+body for each of the commits. As the change is relatively contained, I think maybe we can just squash them into one. Thanks again for working on this!

**Review by Roasbeef — APPROVED** [review]
LGTM 🧬

Changed looking good to me, final nit is just to squash into a single commit. Will test this out a bit locally and also wait on the other reviewer before merging.

**Comment by calvinrzachman** [comment]
@Roasbeef Squashed ✅ Thanks for the guidance on the way through. It was fun to start learning the 2nd most important codebase in the world. I'll be on the lookout for another issue/PR soon.

**Review by cfromknecht — APPROVED** [review]
Thanks @calvinrzachman!! Very clean PR, LGTM 👌

**Review by halseth — APPROVED** [review]
